### ISSUE TITLE
SQLite's `select-tuple` doesn't throw `sql-table-missing`

### ISSUE THREAD

**Issue by AlexIljin** [factor/factor#2081]
Unit-test to reproduce the issue:
```
[
    TUPLE: no-table name ;
    no-table "NO_TABLE" { { "name" "NAME" VARCHAR } } define-persistent
    "none.db" <sqlite-db> [ no-table new select-tuple ] with-db
] [ sql-table-missing? ] must-fail-with
```
The code should throw the `sql-table-missing`, but instead it throws `sqlite-sql-error`. The parsing of this error text is implemented in `parse-sqlite-sql-error`, which is called from `M: sqlite-db-connection parse-db-error`, but it seems that `select-tuple` doesn't go through `execute-statement*`.

**Comment by mrjbq7** [comment]
This fixes it, although in some cases of unable to parse it returns a ``unparsed-sqlite-error`` which is not a ``sql-error``, so I'm not sure if that's groovy or not.

Also, it would be good to add your test case as part of a PR.

Maybe ``unparsed-sqlite-error`` should be a ``sql-error``...?

```diff
diff --git a/basis/db/sqlite/lib/lib.factor b/basis/db/sqlite/lib/lib.factor
index edc8a39f9e..00babd6f4c 100644
--- a/basis/db/sqlite/lib/lib.factor
+++ b/basis/db/sqlite/lib/lib.factor
@@ -1,20 +1,20 @@
 ! Copyright (C) 2008 Chris Double, Doug Coleman.
 ! See http://factorcode.org/license.txt for BSD license.
 USING: accessors alien.c-types alien.data arrays calendar.format
-calendar.parser combinators db db.errors db.sqlite.ffi db.types
-io.backend io.encodings.string io.encodings.utf8 kernel math
-namespaces present sequences serialize urls ;
+calendar.parser combinators db db.errors db.sqlite.errors
+db.sqlite.ffi db.types io.backend io.encodings.string
+io.encodings.utf8 kernel math namespaces present sequences
+serialize urls ;
 IN: db.sqlite.lib
 
 ERROR: sqlite-error < db-error n string ;
-ERROR: sqlite-sql-error < sql-error n string ;
 
 : sqlite-other-error ( n -- * )
     dup sqlite-error-messages nth sqlite-error ;
 
 : sqlite-statement-error ( -- * )
-    SQLITE_ERROR
-    db-connection get handle>> sqlite3_errmsg sqlite-sql-error ;
+    db-connection get handle>> sqlite3_errmsg
+    parse-sqlite-sql-error throw ;
 
 : sqlite-check-result ( n -- )
     {
```

**Comment by AlexIljin** [comment]
Thank you for the quick fix! I thought I had to make everything pass through `execute-statement*` somehow.

Honestly, I don't have a firm opinion about the error class hierarchy. What you are suggesting does make sense.

In my immediate use case I simply wanted to call `[ select-tuple ] ignore-table-missing`, and was surprised it didn't work.

Do you want me to make a PR with the unit-test?

**Comment by mrjbq7** [comment]
Sure if you have time for a PR, otherwise I can do it tonight or tomorrow,
maybe.

**Comment by AlexIljin** [comment]
Unfortunately, removing the `sqlite-sql-error` breaks some `db.sqlite` unit-tests, because for functions like `ensure-table` the code in `M: sqlite-db-connection parse-db-error` takes `n>>` before it passes the string to `parse-sqlite-sql-error` (only if n = `CONSTANT: SQLITE_ERROR 1`).

So, by fixing the methods that don't go through `execute-statement*`+`parse-db-error` we broke the ones that do.

**Comment by mrjbq7** [comment]
No worries. I can poke at it later and clean that up.

**Comment by mrjbq7** [comment]
Fixed in f8de0c1caa72f58fa56ac37e265680086bd1dbf4.
